compute/oslogin/service_account_ssh.py

```diff
@@ -0,0 +1,164 @@
+        logging.info(output)
+    return returncode, output
+# [END run_command_local]
+
+
+# [START create_key]
+def create_ssh_key(oslogin, account, private_key_file=None, expire_time=300):
+    """Generate an SSH key pair and apply it to the specified account."""
+    private_key_file = private_key_file or '/tmp/key-' + str(uuid.uuid4())
+    execute(['ssh-keygen', '-t', 'rsa', '-N', '', '-f', private_key_file])
+
+    with open(private_key_file + '.pub', 'r') as original:
+        public_key = original.read().strip()
+
+    # Expiration time is in microseconds.
+    expiration = int((time.time() + expire_time) * 1000000)
+
+    body = {
+        'key': public_key,
+        'expirationTimeUsec': expiration,
+    }
+    oslogin.users().importSshPublicKey(parent=account, body=body).execute()
+    return private_key_file
+# [END create_key]
+
+
+# [START run_command_remote]
+def run_ssh(cmd, private_key_file, username, hostname):
+    """Run a command on a remote system."""
+    ssh_command = [
+        'ssh', '-i', private_key_file, '-o', 'StrictHostKeyChecking=no',
+        '{username}@{hostname}'.format(username=username, hostname=hostname),
+        cmd,
+    ]
+    ssh = subprocess.Popen(
+        ssh_command, shell=False, stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE)
+    result = ssh.stdout.readlines()
+    if result:
+        return result
+    else:
+        error = ssh.stderr.readlines()
+        return error
+
+# [END run_command_remote]
+
+
+# [START main]
+def main(cmd, project, instance, zone, oslogin=None,
+         account=None, hostname=None):
+    """Run a command on a remote system."""
+
+    # Create the OS Login API object.
+    oslogin = oslogin or googleapiclient.discovery.build('oslogin', 'v1')
+
+    # Identify the service account ID if it is not already provided.
+    account = account or 'users/' + requests.get(SERVICE_ACCOUNT_METADATA_URL,
+                                                 headers=HEADERS).text
+
+    # Create a new SSH key pair and associate it with the service account.
+    private_key_file = create_ssh_key(oslogin, account)
+
+    # Using the OS Login API, get the POSIX user name from the login profile
+    # for the service account.
+    profile = oslogin.users().getLoginProfile(name=account).execute()
+    username = profile.get('posixAccounts')[0].get('username')
+
+    # Create the hostname of the target instance using the instance name,
+    # the zone where the instance is located, and the project that owns the
+    # instance.
+    hostname = hostname or '{instance}.{zone}.c.{project}.internal'.format(
+        instance=instance, zone=zone, project=project)
+
+    # Run a command on the remote instance over SSH.
+    result = run_ssh(cmd, private_key_file, username, hostname)
+
+    # Print the command line output from the remote instance.
+    # Use .rstrip() rather than end='' for Python 2 compatability.
+    for line in result:
+        print(line.decode('utf-8').rstrip("\n\r"))
```

Comment:
You still have double quotes around the `"\n\r"` and low around `".pub"`. There's also a few variables in your test with double quotes that should be single.

Reply:
Done.